pkg/machines/components/createDialog.jsx

```diff
@@ -0,0 +1,376 @@
+            this.onChangedValue(key, e.target.checked)
+        }
+    }
+
+    onChangedValue (key, value) {
+        switch (key) {
+            case 'vendor': {
+                const os = this.props.vendorMap[value][0].shortId;
+                this.setState({
+                    [key]: value,
+                    os,
+                });
+                if (this.props.valuesChanged) {
+                    this.props.valuesChanged('os', os);
+                }
+                break;
+            }
+            case 'os':
+                this.setState({ [key]: value });
+                break;
+            case 'memorySize':
+                this.setState({ [key]: value });
+                value = convertToUnit(value, this.state.memorySizeUnit, units.MiB)
+                break
+            case 'storageSize':
+                this.setState({ [key]: value });
+                value = convertToUnit(value, this.state.storageSizeUnit, units.GiB)
+                break
+            case 'memorySizeUnit':
+                this.setState({ [key]: value });
+                value = convertToUnit(this.state.memorySize, value, units.MiB)
+                key = 'memorySize'
+                break
+            case 'storageSizeUnit':
+                this.setState({ [key]: value });
+                value = convertToUnit(this.state.storageSize, value, units.GiB)
+                key = 'storageSize'
+                break
+            default:
+                break;
+        }
+
+        if (this.props.valuesChanged) {
+            this.props.valuesChanged(key, value);
+        }
+    }
+
+    render () {
+        const vendorSelectEntries = [];
+
+        if (this.props.familyMap[DIVIDER_FAMILY]) {
+            vendorSelectEntries.push((
+                <Select.SelectEntry data={NOT_SPECIFIED} key={NOT_SPECIFIED}>{NOT_SPECIFIED}</Select.SelectEntry>));
+            vendorSelectEntries.push((<Select.SelectDivider/>));
+        }
+
+        this.props.familyList.forEach(({ family, vendors }) => {
+            if (family === DIVIDER_FAMILY) {
+                return;
+            }
+            vendorSelectEntries.push((<Select.SelectHeader>{family}</Select.SelectHeader>));
+
+            vendors.forEach((vendor) => {
+                vendorSelectEntries.push((
+                    <Select.SelectEntry data={vendor} key={vendor}>{vendor}</Select.SelectEntry>));
+            });
+        });
+
+        const osEntries = (
+            this.props.vendorMap[this.state.vendor]
+                .map(os => (<Select.SelectEntry data={os.shortId}
+                                                key={os.shortId}>{getOSStringRepresentation(os)}</Select.SelectEntry>))
+        )
+
+        const unitEntries = ([
+            <Select.SelectEntry data={units.MiB.name} key={units.MiB.name}>{_("MiB")}</Select.SelectEntry>,
+            <Select.SelectEntry data={units.GiB.name} key={units.GiB.name}>{_("GiB")}</Select.SelectEntry>,
+        ])
+
+        return (
```

Comment:
Is it possible to split this huge code into subcomponents?

Reply:
Refactored into MemorySelectRow, but the other fields are left same. I will maybe refactor it again together with the file select.